src/main/java/com/aws/greengrass/mqttclient/MqttClient.java

```diff
@@ -433,50 +452,70 @@ public synchronized void unsubscribe(UnsubscribeRequest request)
-            List<CompletableFuture<?>> publishRequests = new ArrayList<>();
-            while (!Thread.currentThread().isInterrupted() && mqttOnline.get() && spool.getCurrentMessageCount() > 0) {
-                final long id = spool.popId();
-                PublishRequest request = spool.getMessageById(id);
-                if (request == null) {
-                    continue;
-                }
-
-                MqttMessage m = new MqttMessage(request.getTopic(), request.getPayload());
-                publishRequests.add(getConnection(false).publish(m, request.getQos(), request.isRetain())
-                                .whenComplete((packetId, throwable) -> {
-                    // packetId is the SDK assigned ID. Ignore this and instead use the spooler ID
-                    if (throwable == null) {
-                        spool.removeMessageById(id);
-                        logger.atDebug().kv("id", id).kv("topic", request.getTopic())
-                            .log("Successfully published message");
-                    } else {
-                        spool.addId(id);
-                        logger.atError().log("Failed to publish the message via Spooler", throwable);
+    @SuppressWarnings("PMD.AvoidCatchingThrowable")
+    protected void runSpooler() {
+        List<CompletableFuture<?>> publishRequests = new ArrayList<>();
+        while (!Thread.currentThread().isInterrupted()) {
+            try {
+                getConnection(false).connect().get();
+                while (mqttOnline.get()) {
+                    CompletableFuture<Integer> future = publishSingleSpoolerMessage();
+                    if (future == null) {
+                        continue;
                     }
-                }));
+                    publishRequests.add(future);
 
-                if (publishRequests.size() >= maxInFlightPublishes) {
-                    CompletableFuture.anyOf(
-                            publishRequests.toArray(new CompletableFuture[0])).get();
-                    publishRequests = publishRequests.stream().filter(f -> !f.isDone()).collect(Collectors.toList());
+                    if (publishRequests.size() >= maxInFlightPublishes) {
+                        CompletableFuture.anyOf(publishRequests.toArray(new CompletableFuture[0])).get();
+                        publishRequests =
+                                publishRequests.stream().filter(f -> !f.isDone()).collect(Collectors.toList());
+                    }
+                }
+                break;
+            } catch (ExecutionException e) {
+                logger.atError().log("Error when publishing from spooler", e);
+                if (Utils.getUltimateCause(e) instanceof InterruptedException) {
+                    logger.atDebug().log("Shutting down spooler task");
+                    break;
                 }
+            } catch (InterruptedException e) {
+                logger.atDebug().log("Shutting down spooler task");
```

Comment:
I would prefer to log `InterruptedException` at Info or Warn. Customers would want to know if the spooler is shut down. Same for line511.